pkg/defaults/defaults.go

```diff
@@ -218,23 +217,14 @@ func fromConfig(
 						addProvidesForStep(s, params)
 					}
 					imageStepLinks = append(imageStepLinks, snapshot.Creates()...)
-				case resolveConfig.Candidate != nil:
-					value, err = candidate.ResolvePullSpec(httpClient, *resolveConfig.Candidate)
-				case resolveConfig.Release != nil:
-					value, _, err = official.ResolvePullSpecAndVersion(httpClient, *resolveConfig.Release)
-				case resolveConfig.Prerelease != nil:
-					value, err = prerelease.ResolvePullSpec(httpClient, *resolveConfig.Prerelease)
+					continue
+				default:
+					source = releasesteps.NewReleaseSourceFromConfig(resolveConfig, httpClient)
 				}
-				if err != nil {
-					return nil, nil, results.ForReason("resolving_release").ForError(fmt.Errorf("failed to resolve release %s: %w", resolveConfig.Name, err))
-				}
-			}
-			if value != "" {
-				logrus.Infof("Resolved release %s to %s", resolveConfig.Name, value)
-				step := releasesteps.ImportReleaseStep(resolveConfig.Name, nodeName, resolveConfig.TargetName(), value, false, config.Resources, podClient, jobSpec, pullSecret, overrideCLIReleaseExtractImage)
-				buildSteps = append(buildSteps, step)
-				addProvidesForStep(step, params)
 			}
+			step := releasesteps.ImportReleaseStep(resolveConfig.Name, nodeName, resolveConfig.TargetName(), source, false, config.Resources, podClient, jobSpec, pullSecret, overrideCLIReleaseExtractImage)
+			buildSteps = append(buildSteps, step)
+			addProvidesForStep(step, params)
```

Comment:
`addProvidesForStep` has been moved out of the branch. Does it make any difference?

Reply:
Good eye.  The condition (`value == ""`) which caused it and the rest of the block to not be executed was only ever `true` when:

- an integration stream is used as input; we now have a `continue` for that case
- a `RELEASE_*` environment variable was set but empty, in which case you get what you deserve (and an indication in the output logs, because of our unbounded mercy)
- an invalid `resolveConfig`; impossible by prior validation, but we now also detect that in `pkg/steps/release/source.go`

You can argue that the last two cases are a change in behavior, but they are edge cases that I didn't really care to preserve.